### PR TITLE
Heterogeneous graph and new associated RL pipeline.

### PR DESCRIPTION
**Pull request by gostreap**
The purpose of this pull request is to allow the implementation of the new graph convolution discussed in issue #225.

The major novelties are :
- Introduction of a heterogeneous graph with 3 distinct node types for constraints, variables and values.
- The modification of the tripartite graph to better manage `VarViews`:
   `VarView` relations were formerly represented as an edge between 2 distinct variables (the parent and its child). With this PR, `VarView` relations are now represented as constraints of type `ViewConstraint` linked to these 2 variables. When `constraint_type` is set to `true` in `chosen_features`, and a problem containts `VarView` relations, it now encodes the type of the constraint with 4 new dimensions reflecting the exact nature of the `VarView` relation. Please refer to `featurize` function in `heterogeneousstaterepresentation.jl` for more information.
- The implementation of the new convolutions as described in #225.
  - The initialization layer of the convolution :  `HeterogeneousGraphConvInit`
  - The heterogeneous version of `GraphConv`: `HeterogeneousGraphConv`.
- The compatibility of all downstream neural networks (CPNNs) with this new heterogeneous pipeline: `HeterogeneousCPNN`, `HeterogeneousFullFeaturedCPNN` and `HeterogeneousVariableOutputCPNN`.
- The resolution of major bugs in the pipeline:
  - Fix of model emptying after the resolution of each instance
  - Fix of value ordering in the graph construction
- The implementation of a new `mean` aggregation in the convolutional layers.

Minor updates also concern:
- The removal of symmetries in the generation of graphcoloring problems
- The correction of errors in the generation of KEP and RB problems

### CODE REVIEW

**Comment by qcappart** [comment]
Great work ! I am very impressed by the rapidity to implement this. I am very eager to see the results ! :-)

**Comment by gostreap** [comment]
It is not yet fully functional, but we are opening a draft to allow everyone to see the progress and to make comment ! :)

**Comment by louis-gautier** [comment]
The heterogeneous graph pipeline is now running without bugs, but we still have to write some unit tests to ensure that its behavior is as expected.

File: src/RL/representation/default/cp_layer/types.jl
Comment (3rdCore) [reply src/RL/representation/default/cp_layer/types.jl:127]: I think I got a better understanding of `viewConstraints`. These are a type of constraints to specify that a variable is a view (of any kind) of another variable ?

**Comment by 3rdCore** [comment]
Implemented `HeterogeneousVariableOutputCPNN`. I'll add `HeterogeneousFullFeaturedCPNN` on monday. I added the field `possibleValuesIdx` to the `DefaultTrajectoryState` to match `HeterogenousTrajectoryState` fields. 

Many more testsets will be added next week to ensure that the field `possibleValuesIdx` is working as expected.

File: src/RL/representation/default/heterogeneousstaterepresentation.jl
Comment (3rdCore) [inline src/RL/representation/default/heterogeneousstaterepresentation.jl:96]: I definitely need your approval on this line.

Comment (louis-gautier) [reply src/RL/representation/default/heterogeneousstaterepresentation.jl:96]: This line is correct! However, some errors are raised when we compute `valueFeatures` from `possibleValuesIdx` in the case of a `BatchedHeterogeneousTrajectoryState`. I am currently working on it.

File: src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl
Comment (3rdCore) [inline src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl:38]: I know that this hasn't been reconsidered for a while but I don't understand why we have to transform the Cartesian Index into a 1-row matrix of type : `1×4 Matrix{CartesianIndex{2}}`.

Comment (3rdCore) [reply src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl:38]: Well, it is just a matter of dimensions :

Here is a little dummy example done on a handcrafted `batchedDefaultTrajectoryState `: 
```julia
julia> idx = CartesianIndex.(zip([3,4], 1:2))
2-element Vector{CartesianIndex{2}}:
 CartesianIndex(3, 1)
 CartesianIndex(4, 2)

julia> idx_unsqueeze = Flux.unsqueeze(CartesianIndex.(zip([3,4], 1:2)),1)
1×2 Matrix{CartesianIndex{2}}:
 CartesianIndex(3, 1)  CartesianIndex(4, 2)

julia> batched_state_2.fg.nf[:,idx]
3×2 Matrix{Float32}:
 0.0  0.0
 1.0  1.0
 0.0  0.0

julia> batched_state_2.fg.nf[:,idx_unsqueeze]
3×1×2 Array{Float32, 3}:
[:, :, 1] =
 0.0
 1.0
 0.0

[:, :, 2] =
 0.0
 1.0
 0.0
```

**Review by 3rdCore — CHANGES_REQUESTED** [review]

File: src/RL/representation/default/defaultstaterepresentation.jl
Comment (3rdCore) [inline src/RL/representation/default/defaultstaterepresentation.jl:116]: I had to remove my change because it wasn't working anymore with `variableoutputcpnn` (index issue raised at [line 32](https://github.com/corail-research/SeaPearl.jl/blob/heterogeneous/src/RL/nn_structures/variableoutputcpnn.jl)).

I modified the function `from_order_to_id` consequently.

File: src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl
Comment (louis-gautier) [reply src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl:41]: We could potentially try it but I think we should limit the number of weights in our neural networks:  `HeterogeneousGraphConv` already deals differently with variables and values, so it might be unnecessary to further maintain differences in treatment.

File: src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl
Comment (louis-gautier) [reply src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl:38]: Thank you for this example!

File: src/RL/representation/default/defaultstaterepresentation.jl
Comment (louis-gautier) [reply src/RL/representation/default/defaultstaterepresentation.jl:116]: Ok, no problem if it is adapted in `from_order_to_id`

File: src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl
Comment (3rdCore) [reply src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl:41]: Well, the pipeline must be heterogeneous at every step, otherwise, all the work done to increase representation on the convolution will be bottlenecked by this inconsistent layer.  Let's discuss that with the others !

@gostreap @malikattalah @ziadelassal

Comment (3rdCore) [reply src/RL/nn_structures/heterogeneousfullfeaturedcpnn.jl:41]: I have another comment, what if the dimensions of the variable embedding and the value embedding do not match? This requires that the GNN satisfy this condition.

**Comment by louis-gautier** [comment]
> Nice bug spot ! That will indeed simplificate the tripartite graph representation !

I still have to look at the testsets for graph coloring problem generation as tests are now failing.

File: src/RL/representation/default/heterogeneoustrajectorystate.jl
Comment (3rdCore) [inline src/RL/representation/default/heterogeneoustrajectorystate.jl:40]: Well, This is absolutely inefficient and must be removed for production, but I'd prefer having these kinds of verifications to prevent any anti-pattern behavior while trying to make our new pipeline work.

File: src/RL/utils/geometricflux/heterogeneousgraphconvinit.jl
Comment (3rdCore) [inline src/RL/utils/geometricflux/heterogeneousgraphconvinit.jl:11]: The base function is `σ=identity` following what has been done on [this paper.](https://www.sciencedirect.com/science/article/pii/S0952197621004255?via%3Dihub)

File: src/RL/representation/default/heterogeneousstaterepresentation.jl
Comment (3rdCore) [inline src/RL/representation/default/heterogeneousstaterepresentation.jl:245]: @louis-gautier, this 4 lines are very close from line 246 to 249, I'm wondering if this is a copy/paste mistake ?

Comment (3rdCore) [reply src/RL/representation/default/heterogeneousstaterepresentation.jl]: It was.

File: src/CP/constraints/constraints.jl
Comment (gostreap) [reply src/CP/constraints/constraints.jl:74]: The role of this structure is to allow to represent in the graph a link between a variable and associated view.

For example, if we have a `VarView` `y = x + 3`, in the graph we will have two variable nodes which will be linked together by a `ViewConstraint`.

Previously there was an edge between the two variables, which caused several problems.

File: src/RL/utils/geometricflux/graphconv.jl
Comment (gostreap) [inline src/RL/utils/geometricflux/graphconv.jl:23]: The convolutions should be compatible with sum pooling. Eventually, passing through parametric types with different functions would be a solution.

That is, write two functions that are more or less typed like this:
- `function (g::GraphConv{A, B, meanPooling})(fgs::BatchedFeaturedGraph{Float32})`
- `function (g::GraphConv{A, B, sumPooling})(fgs::BatchedFeaturedGraph{Float32})`

This would avoid ifs in the function, which I think is not easy to handle with Zygote.

**Review by gostreap — COMMENTED** [review]
In the current state the convolutions with mean pooling do not work. This must be fixed absolutely before the merge.

**Comment by gostreap** [comment]
Pooling should work now. For `meanPooling`, the `Zygote.ignore()` covers a lot of things, we'll have to watch that it doesn't break the backpropagation.

@3rdCore, can you take a look at my modifications?

File: src/RL/representation/default/defaultstaterepresentation.jl
Comment (louis-gautier) [reply src/RL/representation/default/defaultstaterepresentation.jl]: We don't have any control on the values that are assigned to variables. That's why `variable_assigned_value` always has to be used with `variable_is_bound`, such as the agent can differentiate a situation where the variable is assigned to 0 and a situation where the variable has not been assigned yet.

File: src/RL/utils/geometricflux/heterogeneousgraphconv.jl
Comment (louis-gautier) [reply src/RL/utils/geometricflux/heterogeneousgraphconv.jl:79]: We should not forget to correct this before merging with master. As the function is not recursive, it is ok to return under `Zygote.ignore()` statement but the operations that are computed inside the return here should be copied outside thr `Zygote.ignore()` domain. We have solved the same issue on `heterogeneousgraphtransformer.jl`.

Comment (louis-gautier) [reply src/RL/utils/geometricflux/heterogeneousgraphconv.jl:79]: What causes the Zygote errors is the construction of the `BatchedHeterogeneousFeaturedGraph`. It would even be safer to do separate construction of the object under Zygote.ignore() and return outside it.

File: src/RL/representation/default/cp_layer/types.jl
Comment (3rdCore) [inline src/RL/representation/default/cp_layer/types.jl:84]: can you explain briefly what the issue was with the previous pipeline ?

Comment (louis-gautier) [reply src/RL/representation/default/cp_layer/types.jl:84]: In short, the ordering of the actions taken by the agent was not consistent with the ordering of value node features in the graph. As a result, the values embeddings in `FullFeaturedCPNN` didn't match the actions taken by the agent afterwards. But this fix doesn't impact other CPNNs.

File: src/CP/core/model.jl
Comment (louis-gautier) [inline src/CP/core/model.jl:261]: These lines fix a major bug caused by the fact that the model was not properly emptied after the resolution of each instance in the training process.

File: src/CP/core/model.jl
Comment (louis-gautier) [inline src/CP/core/model.jl:388]: We now take all variables, including `VarViews` into consideration in `global_domain_cardinality`

File: src/RL/representation/default/defaultstaterepresentation.jl
Comment (louis-gautier) [inline src/RL/representation/default/defaultstaterepresentation.jl:173]: If the problem we are trying to solve has `ViewConstraints`, in `constraint_type`, we encode them thanks to 4 new components, respectively: one-hot encoding of the fact the considered constraint is a `ViewConstraint`, multiplication coefficient if the `ViewConstraint` is of type `IntVarViewMul` or `IntVarViewOpposite`, offset coefficient if the if the `ViewConstraint` is of type `IntVarViewOffset`, and one-hot encoding of the fact the `ViewConstraint` is of type `BoolVarViewNot`.

File: src/datagen/coloring.jl
Comment (louis-gautier) [inline src/datagen/coloring.jl:98]: This helps to remove symmetries when generating graphcoloring instances.

**Review by louis-gautier — CHANGES_REQUESTED** [review]
We just have to check that `Zygote.ignore()` statements in all convolutional layers are well positioned and we can merge this branch on `master`.

File: src/CP/core/model.jl
Comment (3rdCore) [reply src/CP/core/model.jl:261]: Nice spot !

File: src/RL/utils/geometricflux/graphconv.jl
Comment (3rdCore) [inline src/RL/utils/geometricflux/graphconv.jl:48]: What is the reason for adding a `replace( M, 0=>1)` ?

Comment (3rdCore) [reply src/RL/utils/geometricflux/graphconv.jl:48]: edit : This is to avoid any division by 0, in case a node has no neighbors, which can happen for value that belongs to no domain after pruning.

Comment (gostreap) [reply src/RL/utils/geometricflux/graphconv.jl:48]: In the case where a node has no neighbor, the sum is zero, which will then cause problems in the division.

Putting a 1 in the division should not be a problem, because it will divide a vector composed only of 0. Any constant other than 1 would have done the job as well.

File: test/datagen/coloring.jl
Comment (louis-gautier) [reply test/datagen/coloring.jl]: Update: I tried different versions for this test but the number of constraints is never consistent. I will keep this test at its previous form.

**Review by louis-gautier — APPROVED** [review]
I think we are now ready to merge with master.

File: src/CP/valueselection/learning/utils.jl
Comment (3rdCore) [inline src/CP/valueselection/learning/utils.jl:200]:
```suggestion
        if isa(state,HeterogeneousTrajectoryState)
```

File: src/RL/representation/default/defaultstaterepresentation.jl
Comment (3rdCore) [inline src/RL/representation/default/defaultstaterepresentation.jl:116]: Did we specifically test this change ?

File: src/RL/representation/default/cp_layer/accessors.jl
Comment (3rdCore) [inline src/RL/representation/default/cp_layer/accessors.jl:172]: Did we tested this ?

File: src/RL/representation/default/defaultstaterepresentation.jl
Comment (gostreap) [reply src/RL/representation/default/defaultstaterepresentation.jl:116]: I don't think we tested it. Maybe we should create an issue for each behavior we want to test, without postponing the merge?

File: src/RL/representation/default/cp_layer/accessors.jl
Comment (louis-gautier) [reply src/RL/representation/default/cp_layer/accessors.jl:172]: Yes, we have unit tests for this in test/heterogeneousstaterepresentation.jl.

**Pull request merged**
